[PAC][AArch64] Lower ptrauth constants in code

kovdan01 opened the pull request:
Depends on #94240.

Define the following pseudos for lowering ptrauth constants in code:

- non-`extern_weak`:
  - no GOT load needed: `MOVaddrPAC` - similar to `MOVaddr`, with added PAC;
  - GOT load needed: `LOADgotPAC` - similar to `LOADgot`, with added PAC;
- `extern_weak`: `LOADauthptrstatic` - similar to `LOADgot`, but use a special stub slot named `sym$auth_ptr$key$disc` filled by dynamic linker during relocation resolving instead of a GOT slot.

Co-authored-by: Ahmed Bougacha <ahmed@bougacha.org>

tschuett commented on llvm/include/llvm/Target/GenericOpcodes.td:
```diff
@@ -110,6 +110,12 @@ def G_GLOBAL_VALUE : GenericInstruction {
   let hasSideEffects = false;
 }
 
+def G_PTRAUTH_GLOBAL_VALUE : GenericInstruction {
```
Could you please add an entry to:
https://github.com/llvm/llvm-project/blob/main/llvm/docs/GlobalISel/GenericOpcode.rst

kovdan01 replied:
Thanks for suggestion! Done

tschuett commented:
You added a new GlobalIsel named opcode. You may want to check it's invariants in the MachineVerifier:
https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/MachineVerifier.cpp

kovdan01 committed:
[PAC][AArch64] Lower ptrauth constants in code

Define the following pseudos for lowering ptrauth constants in code:

- non-`extern_weak`:
  - no GOT load needed: `MOVaddrPAC` - similar to `MOVaddr`, with added PAC;
  - GOT load needed: `LOADgotPAC` - similar to `LOADgot`, with added PAC;
- `extern_weak`: `LOADauthptrstatic` - similar to `LOADgot`, but use a special
  stub slot named `sym$auth_ptr$key$disc` filled by dynamic linker during
  relocation resolving instead of a GOT slot.

Co-authored-by: Ahmed Bougacha <ahmed@bougacha.org>

kovdan01 commented:
> You added a new GlobalIsel named opcode. You may want to check it's invariants in the MachineVerifier: https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/MachineVerifier.cpp

Thanks for suggestion! I've added some basic invariant checking.

kovdan01 commented:
@ahmedbougacha Please note that changes in this PR are a bit different from the https://github.com/ahmedbougacha/llvm-project/tree/eng/arm64e-upstream-llvmorg branch. We ship less changes since some functionality was considered as not needed for ELF. Here are some of the differences.

- There is no support for static materialization of non-`extern_weak` symbols.
- The issue https://github.com/access-softek/llvm-project/issues/79 is fixed. Now not only positive 32 bit offsets are supported properly (while only positive 24 bit ones were supported previously and for 25-32 bit ones there was an assertion failure), but also full-range 64 bit offsets as well (both positive and negative ones).
- Some code related to emitting `$auth_ptr$` stubs is moved to common template helpers which might be used by both `MachineModuleInfoELF` and `MachineModuleInfoMachO`.
- Since `$auth_ptr$` stubs are now only used for `extern_weak` symbols, no offsets are supported in these stubs.

kovdan01 committed:
Fix build

tschuett commented:
In the IRTranslator, you build a G_PTRAUTH_GLOBAL_VALUE by hand. Would an entry to
https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h
simplify the build?

ahmedbougacha commented on llvm/docs/GlobalISel/GenericOpcode.rst:
```diff
@@ -60,6 +60,16 @@ The address of a global value.
+an extra discriminator (64-bit imm).
+
+.. code-block:: none
+  %0:_(p0) = G_PTRAUTH_GLOBAL_VALUE %1:_(p0), s32, %2:_(p0), s64
```
I think this really does need the extra newline after `.. code-block:: none`

kovdan01 replied:
Fixed, thanks, see 2b12acbde48f534e725a1febf1607178266bb3b1

ahmedbougacha reviewed (approved):
LG from my perspective

kovdan01 committed:
Merge branch 'main' into ptrauth-constants-in-code

kovdan01 committed:
Address review comments
